### DISCUSSION TITLE
Download models from Civitai.com directly to Google Drive

### DISCUSSION THREAD

**Discussion by i-huzaifa-arshad** [TheLastBen/fast-stable-diffusion#2016] [Ideas]
So I have been experimenting with ChatGPT recently and I got this idea. I see a lot of amazing models on Civitai and wanted a way to download them directly to Google Drive. Before that, I had to download them to my pc and then upload them to drive (this took like forever). Anyways, I worked around and got this script. I added some models that I found amazing. You can add your favorite model url and name (like Dreamshaper.safetensors) in input field and click add file. It will temporarily add it and show it. Select the checkbox, click download and it will automatically download it in the respective folder. However, there are some issues, this is a temporary thing. Next time you run the cell, it is gone. So, you have to add it again. (Another way to make it permanent is you can add manually in the lists, which I find irritating). I couldn't figure it out how it can be made permanent. With that, I mean when user add model url and name, it should be added to file name and file url list automatically. Hope TheLastBen can fix it. It's a cool feature for me, and hopefully for many others. Here is the code:

```
#@markdown # Civitai Models
#@markdown - Download models directly to google drive

def add_files(btn):
    file_name = file_name_input.value
    file_url = file_url_input.value
    if file_name and file_url:
        file_names.append(file_name)
        file_urls.append(file_url)
        checkboxes.append(widgets.Checkbox(value=False, description=file_name))
        clear_output()
        display_files()
        file_name_input.value = ''
        file_url_input.value = ''

def display_files():
    num_cols = 4
    num_rows = len(checkboxes) // num_cols + (len(checkboxes) % num_cols > 0)
    cols = [widgets.VBox(checkboxes[i*num_rows:(i+1)*num_rows]) for i in range(num_cols)]
    display(widgets.HBox(cols), file_name_input, file_url_input, add_file_button, download_button)

def download_files(btn):
    selected_files = []
    selected_file_names = []
    for i in range(len(checkboxes)):
        if checkboxes[i].value:
            selected_files.append(file_urls[i])
            selected_file_names.append(file_names[i])
            checkboxes[i].description = f"{file_names[i]} selected"
            checkboxes[i].disabled = True
    if len(selected_files) > 0:
        output_path = "/content/gdrive/MyDrive/sd/stable-diffusion-webui/models/Stable-diffusion"
        os.makedirs(output_path, exist_ok=True)
        for file_url, file_name in zip(selected_files, selected_file_names):
            file_path = os.path.join(output_path, file_name)
            !wget -O "$file_path" "$file_url" 
            clear_output()
            inf('Downloaded \u2714','success', '50px')

# List of file URLs
file_urls = [
    "https://civitai.com/api/download/models/15236",
    "https://civitai.com/api/download/models/24365",
    "https://civitai.com/api/download/models/11745",
    "https://civitai.com/api/download/models/22031",
    "https://civitai.com/api/download/models/38904",
    "https://civitai.com/api/download/models/39386"
]

# List of file names
file_names = [
    "Deliberate.safetensors",
    "DreamShaper.safetensors",
    "ChilloutMix.safetensors",
    "Cartoonish.safetensors",
    "TFM-American-Cartoons.safetensors",
    "CartoonStyleClassic.safetensors"
]

# Create the checkboxes
checkboxes = [widgets.Checkbox(value=False, description=f) for f in file_names]

# Create the input fields and buttons
file_name_input = widgets.Text(description='File Name:')
file_url_input = widgets.Text(description='File URL:')
add_file_button = widgets.Button(description='Add File')
add_file_button.on_click(add_files)

# Create the download button
download_button = widgets.Button(description="Download Model(s)")
download_button.on_click(download_files)

# Display the checkboxes and the download button in four columns
display_files()
```

**Comment by TheLastBen** [comment]
you can use the "model download" cell to download any model directly to your gdrive, it is all integrated in the notebook

**Reply by i-huzaifa-arshad** [reply]
Yes, you are right. We can use that cell. However, it doesn't work properly. If you paste the civitai link, it will start downloading. But it will save the model by the wrong name (model.safetensors).

![1](https://user-images.githubusercontent.com/76521534/233641133-48b47243-797a-454e-ad96-84ec55d0bab8.jpg)

Also, if you check your Google Drive folder, the default model.ckpt is replaced by this model.safetensors, which is not good as we lost the default SD model. And the model now present is the one we downloaded from civitai.

![2](https://user-images.githubusercontent.com/76521534/233641869-2ab06c47-8753-4468-b420-5b29274ae071.jpg)

With the code I provided, you have the option to rename the model and it won't affect the default model. Please let me know your thoughts.

**Reply by alperinugur** [reply]
good stuff. I used  "!wget -O myfile.ext http_Path_to_Civit_File.ext"  which also works.

Actually if you insert the -b option, the download process will be in background, so you could enjoy other models while it is downloading

**Reply by ackvf** [reply]
You can also use `gdown`

```
!gdown -O $target "$url"
```

**Comment by TheLastBen** [comment]
I did that on purpose to avoid filling up gdrive space

**Reply by i-huzaifa-arshad** [reply]
I see. Well, you can modify your code to save the model by its own name or the name of user's choice. Also, you can add the option to deleted the unwanted models if space is the issue. If someone needed the model again, they simply can redownload it.

**Comment by ackvf** [comment]
I have another solution for that that is far more elegant, have a look here! https://github.com/TheLastBen/fast-stable-diffusion/discussions/2018

It is using this colab notebook
https://colab.research.google.com/drive/1eZbKtuP4XbC6wOfxmu_2_UtjLmbq0Z6a?usp=sharing

![image](https://user-images.githubusercontent.com/2313018/233853126-853f414c-6def-401c-8399-c9bc5ae38636.png)

**Reply by i-huzaifa-arshad** [reply]
Glad to see someone improved my approach. That's exactly what I wanted. My approach was only for models, not for LoRas, hypernetworks, etc. It's great you came up with this notebook where each cell corresponds to different thing. Thank you.

**Reply by i-huzaifa-arshad** [reply]
Also one more thing, you need to modify your code as it gives the warning. Just modify the output path thing as TheLastBen did. Moreover, the hyperlinks (gdrive/Lora, etc) you provided require access, and I am not sure what these links are referring to. Maybe clicking them opens the respective folder in gdrive and this works for you, but not for us.

**Reply by ackvf** [reply]
Oh right, those are just links to my google drive so that I can preview the folder. This is meant to be replaced if you fork the notebook. Do you have a suggestion how to make it better / clear?

BTW I am not sure what you mean by "modify the output path". Never noticed any warnings. Thanks for the input!

**Reply by i-huzaifa-arshad** [reply]
Maybe you can make the hyperlink generic. like " `/content/gdrive/MyDrive/sd/stable-diffusion-webui/models/Stable-diffusion/Lora` ". Not sure if you did the same, but try it. And by modifying the path I meant change the path. TheLastBen did modify the path to avoid google colab warning, see this pic.

![image](https://user-images.githubusercontent.com/76521534/234090930-2afdd6ee-4b74-475b-abc3-9b213d8af835.png)

I guess google has blacklisted some words like "webui" as per their updated rules. TheLastBen modified code like this:

Before: `/content/gdrive/MyDrive/sd/stable-diffusion-webui`
After: `/content/gdrive/$mainpth/sd/stable-diffusion-$blasphemy`

Check the `Install/Update AUTOMATIC1111 repo` cell in latest [notebook](https://github.com/TheLastBen/fast-stable-diffusion/blob/main/fast_stable_diffusion_AUTOMATIC1111.ipynb) for reference for `$mainpth` and `$blasphemy`. Then you will understand it properly.

**Reply by ackvf** [reply]
Oh I see, I was wondering what that change was for.

**Reply by i-huzaifa-arshad** [reply]
Let's hope google devs don't see this post xD Or they will comeup with better ideas to stop SD running on Colab.

**Comment by Mossbraker** [comment]
There's also [this tool](https://github.com/vsp245/Colab-AutoWebUI-Downloader), a 'Download, Upload, File Manager for "AUTO111 SD WebUI" and Google Colab' which you may find to be of use.

**Reply by i-huzaifa-arshad** [reply]
Will look into it.

**Reply by ackvf** [reply]
Oh neat!

**Comment by dilectiogames** [comment]
the simplest way is to use this line of code after you have connected to the drive
`!gdown https://huggingface.co/Norisuke193/kl-f8-anime2/resolve/main/kl-f8-anime2.vae.pt -O kl-f8-anime2.vae.pt`
`!gdown WEBLINK -O NAME_OF_FILE" add the correct name
but you need to move the file where it needs to be. it will be downloaded in the SD main folder or the models main folder. sometimes it will not appear there but it will inside the google notebok file browser (the one in the left) drag and drop it where it needs to be.

why it does one of the above at random is a mystery
